pkg/util/modules/modules.go

```diff
@@ -0,0 +1,153 @@
+package modules
+
+import (
+	"fmt"
+
+	"github.com/cortexproject/cortex/pkg/util/services"
+	"github.com/pkg/errors"
+)
+
+// module is the basic building block of the application
+type module struct {
+	// dependencies of this module
+	deps []string
+
+	// initFn for this module (can return nil)
+	initFn func() (services.Service, error)
+}
+
+// Manager is a component that initialises modules of the application
+// in the right order of dependencies.
+type Manager struct {
+	modules map[string]*module
+}
+
+// NewManager creates a new Manager
+func NewManager() *Manager {
+	return &Manager{
+		modules: make(map[string]*module),
+	}
+}
+
+// RegisterModule registers a new module with name and init function
+// name must be unique to avoid overwriting modules
+// if initFn is nil, the module will not initialise
+func (m *Manager) RegisterModule(name string, initFn func() (services.Service, error)) {
+	m.modules[name] = &module{
+		initFn: initFn,
+	}
+}
+
+// AddDependency adds a dependency from name(source) to dependsOn(targets)
+// An error is returned if the source module name is not found
+func (m *Manager) AddDependency(name string, dependsOn ...string) error {
+	if mod, ok := m.modules[name]; ok {
+		mod.deps = append(mod.deps, dependsOn...)
+	} else {
+		return fmt.Errorf("no such module: %s", name)
+	}
+	return nil
+}
+
+// InitModuleServices initalises the target module by initalising all its dependencies
+// in the right order. Modules are wrapped in such a way that they start after their
+// dependencies have been started and stop before their dependencies are stopped.
+func (m *Manager) InitModuleServices(target string) (map[string]services.Service, error) {
+	servicesMap := map[string]services.Service{}
+
+	// initialize all of our dependencies first
+	deps := m.orderedDeps(target)
+	deps = append(deps, target) // lastly, initialize the requested module
+
+	for ix, n := range deps {
+		mod := m.modules[n]
+
+		var serv services.Service
+
+		if mod.initFn != nil {
+			s, err := mod.initFn()
+			if err != nil {
+				return nil, errors.Wrap(err, fmt.Sprintf("error initialising module: %s", n))
+			}
+
+			invDeps := m.findInverseDependencies(n, deps[ix+1:])
+			if s == nil {
+				if len(invDeps) > 0 {
```

Comment:
If a service returns `nil` like this, but without an error that means we will probably need a stub service instead of throwing an error here. There are some modules that are really only meant for orchestration and don't need a running service.

Reply:
I think such modules should not be used as a dependency for other modules. In that case, error is valid outcome here. WDYT?

Reply:
In that case we would need to rearchitect other modules to return a service. For instance the API module only exists to register routes with the server. It doesn't need service to manage it since it only is required during initialization. Since every service that serves web traffic depends on it, it will need to return a service that does nothing.

Reply:
OK, I see. Let's not treat nil service for module with other modules depending on it as an error then. (It's would be a change from what Cortex is currently doing anyway).